src/main/java/com/lmax/disruptor/BatchEventProcessor.java

```diff
@@ -37,31 +38,112 @@
     private ExceptionHandler<? super T> exceptionHandler;
     private final DataProvider<T> dataProvider;
     private final SequenceBarrier sequenceBarrier;
-    private final EventHandler<? super T> eventHandler;
+    private final BaseEventHandler<? super T> eventHandler;
+    private final int batchLimitOffset;
     private final Sequence sequence = new Sequence(Sequencer.INITIAL_CURSOR_VALUE);
     private BatchRewindStrategy batchRewindStrategy = new SimpleBatchRewindStrategy();
     private int retriesAttempted = 0;
+    private final boolean rewindable;
```

Comment:
Rather than a `boolean`, could we have behaviour?

Reply:
Thinking something like adding an private inner interface:

```java
private final interface RewindBehaviour
{
      long onRewindableException(RewindableException e);
}
```

and then there could be two private methods:

```java
private long beKindRewind(final RewindableException e)
{
    if (this.batchRewindStrategy.handleRewindException(e, ++retriesAttempted) == REWIND)
    {
        return startOfBatchSequence;
    }
    else
    {
        retriesAttempted = 0;
        throw e;
    }
}

private long rewindNotAllowed(final RewindableException e)
{
    throw new RuntimeException("Rewindable Exception thrown from a non-rewindable event handler", e);
}
```

Constructors would then take `this::beKindRewind` in place of `true` and `this::rewindNotAllowed` in place of false? (one catch here is that they might not be referencable in `this()` calls, so eh?).

This would save continuously checking an `if` that will never change; on the other hand it does possibly make navigation slightly harder?

Also, this really only works because `retriesAttempted` appears to be instance scoped? I'm assuming this is so we persist failures across calls until we get a success, but it's not entirely clear to me?